### PR TITLE
[5.7] Support dot notation in Request object when using ArrayAccess

### PR DESCRIPTION
**Pull request by TitasGailius**
Moved from 5.6 to 5.7.
https://github.com/laravel/framework/pull/23453

Currently you can access request's array input using dot notation like this:
```php
$request->input('foo.bar');
// It's also supported when using
$request->only('foo.bar');
$request->all('foo.bar');
```
Although when you try to do the same thing using ArrayAccess (`$request['foo.bar']`) it's not supported which makes it a bit inconsistent.

This adds support of the dot notation when used with ArrayAccess.
```php
$request['foo.bar'];
// and of course
isset($request['foo.bar']);
```

### CODE REVIEW

**Comment by carusogabriel** [comment]
@TitasGailius This is still targeting the `5.6` branch. Please, change to `5.7`.

**Comment by GrahamCampbell** [comment]
@caiquecastro no, it's not?

**Comment by taylorotwell** [comment]
You have a big breaking change in the __get method by using Arr::get... all `null` inputs will now use the "default" closure and fallback to the route variables. That was not the prior behavior.

**Comment by TitasGailius** [comment]
@taylorotwell I tested it and `Arr::get` checks whether key `exists` in the array before calling the "default" callback and doesn't check if it's null. So none of these would trigger the default callback:

```php
Illuminate\Support\Arr::get(['foo' => ['bar' => null]], 'foo.bar', function () {
    dd('wrong');
});
```
```php
Illuminate\Support\Arr::get(['foo' => null, 'foo', function () {
    dd('wrong');
});
```
which makes these 2 identical:
```php
Illuminate\Support\Arr::get(['foo' => null], 'foo', function () {
    $this->route('foo');
});

// identical

if (array_key_exists('foo', ['foo' => null])) {
    return data_get(['foo' => null], 'foo');
}

return $this->route('foo');
```

Or do I miss something here?

File: tests/Http/HttpRequestTest.php [hunk 1]
```diff
@@ -308,6 +308,32 @@ public function testInputMethod()
...
+        $request->setRouteResolver(function () use ($request) {
+            $route = new Route('GET', '/foo/bar/{id}', []);
+            $route->bind($request);
+            $route->setParameter('id', 'foo');
+
+            return $route;
+        });
+
+        $this->assertFalse(isset($request['non-existant']));
+        $this->assertNull($request['non-existant']);
```
Comment (vlakoff) [inline tests/Http/HttpRequestTest.php]: spelling: it's "non-existent" ;)

**Pull request closed without merging**